### PR TITLE
Build a URL Shortener with React, TypeScript, and Directus

### PR DESCRIPTION
**Pull request by xKachi**
Complete draft for the issue  [#156](https://github.com/directus-labs/guest-authoring/issues/156)

Project Link --> https://github.com/xKachi/Link-Shortener

### CODE REVIEW

**Comment by phazonoverload** [comment]
Thanks for your patience. I just wanted to let you know that I'll be reviewing this in the next couple of days.

File: Build a URL Shortener with React, TypeScript, and Directus/index.md.md [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+}
+
+export const directus = createDirectus(directusUrl)
+  .with(staticToken(directusToken))
+  .with(rest());
+```
+
+- `staticToken`: is used to authenticate the client we'll create.
+- `createDirectus`: is a function that initializes a Directus client.
+- `rest()`: enables the `.request(...)` method to query the collection.
```
Comment (phazonoverload) [inline Build a URL Shortener with React, TypeScript, and Directus/index.md.md:201]:
````suggestion
## Setting Up Your React Project

Install React.js, set up dependencies, and run a development server:

```
npm create vite@latest link-shortener
? Select a framework: React
? Select a variant: TypeScript

npm install
npm install --save-dev @types/node
npm install react-router-dom
npm install @directus/sdk 

npm run dev
```

Now, open http://localhost:5173/ on your browser, and you should see the starter page.

Next, we will set up Vite to allow environment variables, which will be used to store our Directus credentials. In the `vite.config.ts` file, add the following code:

```typescript
import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";

export default defineConfig({
  plugins: [react()],
  define: {
    "process.env.VITE_DIRECTUS_API_TOKEN.": JSON.stringify(
      process.env.VITE_DIRECTUS_API_TOKEN
    ),
    "process.env.VITE_DIRECTUS_API_URL.": JSON.stringify(
      process.env.VITE_DIRECTUS_API_URL
    ),
  },
});
```

Create a `.env` file in the `src` directory, and the following variables, being sure to provide your specific static authentication token and project URL:

```
VITE_DIRECTUS_API_TOKEN = XXXXX
VITE__DIRECTUS_API_URL = https://your-amazing.directus.app/
```

In the `src` directory create a sub-directory called `utils`, and within it, a `directus.ts` file.

Add the following code to _directus.ts_:

```typescript
import { createDirectus, staticToken, rest } from "@directus/sdk";

const directusToken = import.meta.env.VITE_DIRECTUS_API_TOKEN;
const directusUrl = import.meta.env.VITE__DIRECTUS_API_URL;

if (!directusToken) {
  throw new Error("Please include a Token");
}

if (!directusUrl) {
  throw new Error("Please include a Url");
}

export const directus = createDirectus(directusUrl)
  .with(staticToken(directusToken))
  .with(rest());
```
````

File: Build a URL Shortener with React, TypeScript, and Directus/index.md.md [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+    </>
+  );
+}
+
+export default App;
+```
+
+Browser Output:
+
+![Dynamic URL Output](<Creating the Dynamic Route (1).gif>)
```
Comment (phazonoverload) [inline Build a URL Shortener with React, TypeScript, and Directus/index.md.md:249]:
````suggestion
### Creating the Dynamic Route

Let's create the route that will be used to redirect the user to the slug URL queried from Directus. In the `App.tsx` file add the following code:

```typescript
import { BrowserRouter, Routes, Route } from "react-router-dom";
import { useParams } from "react-router-dom";

function App() {
  return (
    <>
      <h1>Link Shortener</h1>
      <BrowserRouter>
        <Routes>
          <Route path="/:slug" element={<LinkRoute />}></Route>
        </Routes>
      </BrowserRouter>
    </>
  );
}

function LinkRoute() {
  const { slug } = useParams();

  console.log(slug);

  return (
    <>
      <h1>{slug}</h1>
    </>
  );
}

export default App;
```
````

File: Build a URL Shortener with React, TypeScript, and Directus/index.md.md [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+2. The response data is filtered to find the short link that matches the current slug.
+3. If no matching short link is found, it sets an error message and throws an error.
+4. If a matching short link is found, it updates the short links `click` property by `1` via the `updateItem` composable from Directus API.
+5. Finally, the user is redirected to the URL associated with the short link using the `window.location.assign` method.
+
+Finally, if an error occurs during the data fetching process, the error is caught and logged to the console. The `LinkRoute` component also returns the error message if the slug is invalid.
+
+Browser Output:
+
+![Slug URL Output](<Querying Directus.gif>)
```
Comment (phazonoverload) [inline Build a URL Shortener with React, TypeScript, and Directus/index.md.md:348]: Please can you take this section into smaller, more digestible steps. Please also make use of our diff formatting (check the writing guide) to draw attention to additions/deletions.

File: Build a URL Shortener with React, TypeScript, and Directus/index.md.md [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+## Conclusion
+
+You’ve just learned how to set up the data collection process and implement dynamic routing using React and Directus. The broad process was:
+
+- Set up Directus data model
+- Customize user roles and permissions
+- API endpoint to query data from Directus
+- Dynamic Route to navigate to slug URL
+
+I hope you find this tutorial useful - if you have any questions or hurdles feel free to join the Directus [official Discord server](https://discord.com/invite/directus).
```
Comment (phazonoverload) [inline Build a URL Shortener with React, TypeScript, and Directus/index.md.md:359]:
```suggestion
## Conclusion

You’ve just learned how to set up the data collection process and implement dynamic routing using React and Directus. The broad process involved setting up a Directus data model, customizing user roles and permissions, creating an API endpoint to query data from Directus, and configuring a dynamic route to navigate to slug URLs.

I hope you find this tutorial useful - if you have any questions or hurdles feel free to join the Directus [official Discord server](https://directus.chat).
```

**Review by phazonoverload — CHANGES_REQUESTED** [review]
Good start - I've left lots of comments and suggestions, mostly rooted in keeping it concise. You don't need to teach readers every part of the Directus UI - just how to piece functionality together to build a project.

File: Build a URL Shortener with React, TypeScript, and Directus/index.md.md (hunk 3 above)
Comment (xKachi) [reply Build a URL Shortener with React, TypeScript, and Directus/index.md.md]: Done.

**Pull request closed without merging**